From 01cad28cbaf8d506b96987c34e488c9792582da5 Mon Sep 17 00:00:00 2001
From: Eric Ruei <e-ruei1@ti.com>
Date: Fri, 27 Jan 2017 09:45:39 -0500
Subject: [PATCH] Fix crash in WebEngineContext when using Wayland

For some reason the OpenGL context wayland QPA sets has no nativeHandle,
so we end up crashing in strcmp.

Assume a context without nativeHandle is Wayland or other GLES2 platform
and also force GLES2 when using Ozone.

@note: Backported from QT 5.7/8 release

Change-Id: Ia3fc524f3ffbb278d86f9153ec96c7258ef86656
Reviewed-by: Michal Klocek <michal.klocek@qt.io>
Signed-off-by: Eric Ruei <e-ruei1@ti.com>
---
 src/core/web_engine_context.cpp | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/src/core/web_engine_context.cpp b/src/core/web_engine_context.cpp
index 3289a3c..d6a6307 100644
--- a/src/core/web_engine_context.cpp
+++ b/src/core/web_engine_context.cpp
@@ -265,8 +265,11 @@ WebEngineContext::WebEngineContext()
 
     const char *glType = 0;
     if (!usingANGLE() && !usingSoftwareDynamicGL() && !usingQtQuick2DRenderer()) {
-        if (qt_gl_global_share_context()) {
-            if (!strcmp(qt_gl_global_share_context()->nativeHandle().typeName(), "QEGLNativeContext")) {
+        if (qt_gl_global_share_context() && qt_gl_global_share_context()->isValid()) {
+            if ((qt_gl_global_share_context()->nativeHandle().isNull()) ||
+            // If the native handle is QEGLNativeContext try to use GL ES/2, if there is no native handle
+            // assume we are using wayland and try GL ES/2, and finally Ozone demands GL ES/2 too.
+                (!strcmp(qt_gl_global_share_context()->nativeHandle().typeName(), "QEGLNativeContext"))) {
                 if (qt_gl_global_share_context()->isOpenGLES()) {
                     glType = gfx::kGLImplementationEGLName;
                 } else {
-- 
1.9.1

